Add extractMatches() to RegExUtils

komatzu1337 opened the pull request:
<!--
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
-->

Provides simple way to get all regex matches as list without manual Matcher handling.

garydgregory reviewed (commented):
I'm not sure how this one. This is a pretty narrow use case. Some people will want a List, others an array; some will want String input, others will say it should be a CharSequence. What should happen if the string is null? What should happen if the pattern is null? Should an empty list (or array) be returned? A null? Should the List be immutable? Can this even be reused with Lang?

komatzu1337 commented:
TY for feedback! I implemented the following improvements:

Now throws IllegalArgumentException with clear messages for null 
Supports both String and CharSequence 
Provides both List<String> and String[] return 
Added regex-string variant 
All returned collections are unmodifiable
Never returns null (empty collections for null) 
Improved documentation

I will add tests later if you find this class worthwhile

ppkarwasz commented on src/main/java/org/apache/commons/lang3/RegExUtils.java:
```diff
@@ -751,4 +755,64 @@ private static String toStringOrNull(final CharSequence text) {
     }
+
+
+    public class MatchFinder {
```
There is no need for a nested class.

ppkarwasz commented on src/main/java/org/apache/commons/lang3/RegExUtils.java:
```diff
@@ -751,4 +755,64 @@ private static String toStringOrNull(final CharSequence text) {
+         * @throws IllegalArgumentException if text or pattern is null
+         */
+        public List<String> findMatches(CharSequence text, Pattern pattern) {
+            if (text == null)
+                throw new IllegalArgumentException("Text must not be null");
+
+            if (pattern == null)
+                throw new IllegalArgumentException("Pattern must not be null");
```
Since most of the methods in Commons Lang are `null`-safe, it would be more appropriate to return an empty list.

komatzu1337 replied:
I didn't know that, your suggest seems right

komatzu1337 replied:
```suggestion
         if (text == null || pattern == null){
            return Collections.emptyList();
        }
```

ppkarwasz replied:
Fixed in https://github.com/search?q=findAllMatches+language%3AJava&type=code

ppkarwasz commented on src/main/java/org/apache/commons/lang3/RegExUtils.java:
```diff
@@ -751,4 +755,64 @@ private static String toStringOrNull(final CharSequence text) {
+         * @return a list of found matches
+         * @throws IllegalArgumentException if text or pattern is null
+         */
+        public List<String> findMatches(String text, Pattern pattern) {
+            return findMatches((CharSequence) text, pattern);
+        }
```
Since `String` is a `CharSequence`, this method seems useless.

ppkarwasz reviewed (changes requested):
Hi @komatzu1337,

Thanks for the PR, could you add some unit tests please?

Note also that you [enabled "Vigilant mode"](https://docs.github.com/en/authentication/managing-commit-signature-verification/displaying-verification-statuses-for-all-of-your-commits) on your GitHub account, but you didn't [add the GPG key you are using to your GitHub account](https://docs.github.com/en/authentication/managing-commit-signature-verification/adding-a-gpg-key-to-your-github-account). For this reason all your commits appear as "Unverified". Can you fix that?

garydgregory commented:
I'm still not convinced this belongs here.

ppkarwasz commented:
> I'm still not convinced this belongs here.

@komatzu1337, I appreciate the effort you've put into this contribution. I also have some reservations about the utility of `findMatches`, so please view my comments as suggestions for improvement rather than a final decision on acceptance.

Just as a reminder, per ASF guidelines, all maintainers need to reach consensus before any changes are merged.

ppkarwasz commented:
> > I'm still not convinced this belongs here.
> 
> why?

When adding a new method to Commons Lang, we need to consider the impact on the entire ecosystem. Even small additions increase the library's size for thousands of downstream projects. Given the wide range of potential utility methods, we prioritize those likely to benefit a significant portion of users—typically, at least 1%.

While your proposed method may be useful in some contexts, it might not offer enough broad utility to justify inclusion in the core library.

komatzu1337 commented:
@ppkarwasz

ppkarwasz commented on src/main/java/org/apache/commons/lang3/RegExUtils.java:
```diff
@@ -751,4 +755,67 @@ private static String toStringOrNull(final CharSequence text) {
+     * @param pattern the pattern to search for
+     * @return an unmodifiable list of found matches; returns an empty List if text or pattern is null
+     */
+    public List<String> findMatches(CharSequence text, Pattern pattern) {
```
```suggestion
    public static List<String> findAll(CharSequence text, Pattern pattern) {
```

This is an utility class, so all the methods must be `static`.

_Nitpick_: Naming is certainly on of the hardest tasks. Since all the methods in this class deal with matching and have names like `replaceAll`, `removeAll`, I would call this method `findAll`.

ppkarwasz commented on src/main/java/org/apache/commons/lang3/RegExUtils.java:
```diff
@@ -751,4 +755,67 @@ private static String toStringOrNull(final CharSequence text) {
+     * @param regex the regular expression pattern
+     * @return an array of found matches; returns an empty array if text or regex is null
+     */
+    public String[] findMatchesAsArray(CharSequence text, String regex) {
```
Looking at the remaining methods of this class, each method with signature `(CharSequence, Pattern)` has a corresponding method with signature `(String, String)`. Therefore I would define a method:

```java
public static List<String> findAll(String text, String regex)
```

I don't think it is useful to have both a method returning `List<String>` and `String[]`. The return type should be coherent.

ppkarwasz commented on src/main/java/org/apache/commons/lang3/RegExUtils.java:
```diff
@@ -751,4 +755,67 @@ private static String toStringOrNull(final CharSequence text) {
+     * @param pattern the pattern to search for
+     * @return a modifiable ArrayList of found matches; returns an empty List if text or pattern is null
+     */
+    public List<String> findMatchesModifiable(CharSequence text, Pattern pattern){
```
-1

We should not start multiplying methods.

ppkarwasz commented on src/test/java/org/apache/commons/lang3/RegExUtilsTest.java:
```diff
@@ -379,4 +381,149 @@ public void testReplacePatternDeprecated() {
 
+    private RegExUtils utils;
+
+    @BeforeEach
+    void setUp() {
+        utils = new RegExUtils();
+    }
```
Utility classes are never instantiated.

ppkarwasz commented on src/test/java/org/apache/commons/lang3/RegExUtilsTest.java:
```diff
@@ -379,4 +381,149 @@ public void testReplacePatternDeprecated() {
+    }
+
+    @Test
+    public void testFindMultipleMatches() {
```
Can you use parameterized logging instead of multiple methods with the same logic?

ppkarwasz reviewed (changes requested):
@komatzu1337,

As noted earlier, we're still not convinced that even a single findAll method is broadly useful. Including three separate implementations feels like more than we can justify at this point. Could you please limit the pull request to the `findAll(CharSequence, Pattern)` method, and optionally a second variant that delegates to it?

garydgregory reviewed (commented):
There are so many possibilities of combining `Pattern`, `Matcher`, and `MatchResult` that I don't see why this _specific_ use case needs to be here. For example, where would this be reused in Commons Lang or from other Commons libraries?

ppkarwasz commented:
> There are so many possibilities of combining `Pattern`, `Matcher`, and `MatchResult` that I don't see why this _specific_ use case needs to be here. For example, where would this be reused in Commons Lang or from other Commons libraries?

Good point! @komatzu1337, could you share a few concrete examples where this method would be especially useful—either in real-world projects or potential usage within Commons libraries?

The current unit tests illustrate some interesting scenarios, but they can be addressed using existing utilities:

- Splitting a sentence into words can be done with `StringUtils.split`.
- To extract numbers from a sentence, you could split into words and then filter those that are numeric.

Understanding your intended use cases would help us evaluate whether this method brings enough added value and general applicability to warrant inclusion here.

komatzu1337 commented:
@ppkarwasz @garydgregory Thank you for your patience and valuable feedback.After considering your feedback, i agree that this method doesn’t align well with Commons Lang’s scope. I apologize for the time spent reviewing this proposal - in hindsight, I should have more thoroughly vetted the use case before submitting it. 

This experience has helped me better understand Commons Lang's design philosophy. I will be sure to:
1. Review the contribution guidelines more carefully next time
2. Get feedback in advance through discussion questions for future changes.
